File: agents/livekit/agent-lib/recording/encryption-key.js
import { randomBytes } from 'node:crypto';

export const GCM_IV_LENGTH = 12;
export const GCM_AUTH_TAG_LENGTH = 16;
export const KEY_LENGTH = 32;

/**
 * Derive a 32-byte key from a client-provided string.
 *
 * Clients pick any string they like; we either truncate or right-pad with
 * zeros so the resulting buffer always matches AES-256's key size. The
 * Python sibling implements the same derivation so a recording encrypted on
 * one platform decrypts on the other.
 *
 * @param {string} clientKey
 * @returns {Buffer}
 */
export function deriveKey(clientKey) {
  const keyBytes = Buffer.from(clientKey, 'utf8');
  if (keyBytes.length >= KEY_LENGTH) {
    return keyBytes.subarray(0, KEY_LENGTH);
  }
  return Buffer.concat([keyBytes, Buffer.alloc(KEY_LENGTH - keyBytes.length, 0)]);
}

/**
 * Generate a random 32-byte key plus its base64 representation. The base64
 * value is what the server stores so it can decrypt on the operator's behalf
 * during download. Returned `key` is the raw buffer used for the in-stream
 * cipher.
 *
 * @returns {{ key: Buffer, base64: string }}
 */
export function generateKey() {
  const key = randomBytes(KEY_LENGTH);
  return { key, base64: key.toString('base64') };
}

File: agents/pipecat/pipecat_aplisay/recording/encryption_key.py
"""Key derivation / generation — sibling of ``lib/recording/encryption-key.js``.

Two modes, both producing a 32-byte AES key:

- ``derive_key(client_key)`` — truncate or zero-pad the UTF-8 bytes of a
  client-provided string. Reproduces the JS ``deriveKey`` exactly so a key
  picked on one side decrypts on the other.
- ``generate_key()`` — random 32 bytes, returned alongside its base64 form
  for storage in the call metadata.
"""

from __future__ import annotations

import base64
import secrets

GCM_IV_LENGTH = 12
GCM_AUTH_TAG_LENGTH = 16
KEY_LENGTH = 32


def derive_key(client_key: str) -> bytes:
    """Truncate-or-zero-pad a client string to the AES-256 key size."""
    raw = client_key.encode("utf-8")
    if len(raw) >= KEY_LENGTH:
        return raw[:KEY_LENGTH]
    return raw + b"\x00" * (KEY_LENGTH - len(raw))


def generate_key() -> tuple[bytes, str]:
    """Return ``(raw_key, base64_key)``. The base64 form is what's stored
    server-side so the platform can decrypt on download.
    """
    key = secrets.token_bytes(KEY_LENGTH)
    return key, base64.b64encode(key).decode("ascii")
